### ISSUE TITLE
decouple swoole to core and middleware parts

### ISSUE THREAD

**Issue by jk2K** [swoole/swoole-src#2295]
attempts to split swoole to core and middleware parts, let the core team focus on core functionality development

middleware, such as http-server, redis, mysql, Hbase, ZooKeeper, ElasticSearch, maybe we can use PHP to implement, based on swoole ability, distributed by composer, If we use PHP to develop, the middleware package can get more contributions from the community

core feature, such as Event-Driven, Coroutine, we continue to develop with C/C++, install by PHP extension

This re-organization is also a good practice for maintenance, as these packages will be released separated, thus allowing more frequent updates

I propose to create multi repository, for example

- swoole/core: core feature, develop with C/C++, install by PHP extension
- swoole/swoole-http-server: use PHP, install by composer
- swoole/swoole-mysql: use PHP, install by composer
- swoole/swoole-redis: use PHP, install by composer
- to be continued

**Comment by need-sleep** [comment]
One think to keep in mind is that some of swoole's features are not possible in vanilla PHP or it's mainline extensions. For example PHP's MySQL extension (or any others in fact) lack the concept of coroutines, requiring swoole to maintain it's own. Wouldn't this make separation inconstant?

**Comment by jk2K** [comment]
[PHP RFC: FFI - Foreign Function Interface](https://wiki.php.net/rfc/ffi) has been accepted, may be we can call the inevitable swoole C modules via FFI, to reduce some obstacles

**Comment by need-sleep** [comment]
Something like FFI can be awesome for PHP's future, but the PHP Group is slow at making these sorts of changes because of how complex they can be. The push that they have made with their incorporation of an AST is a big start, but even this is muddy water in areas. In the end the entire extension community will need to add hooks for these features. For example, how will FFI help make the standard MySQLi extension capable of understanding coros if FFI is used to make a native coroutine feature set like Swoole's when it doesn't even understand PHP coroutine's? In other words, FFI will make PHP capable of making native code, but what will it take for existing native extensions to work with these native instructions created by FFI? It will likely prove great for making algorithms for sorting things or even complex algorithms for codecs and ciphering, but even working with basic system libraries such as sockets will prove challenging, especially if your plan is to change how even basic libraries work.

**Comment by need-sleep** [comment]
I hope you don't tire of my 2 cents as I'm not an official voice of Swoole, but I really think Swoole should continue on the path they are going. In the near future it's possible to see other languages/platforms embracing Swoole such as Java and Node.js which can prove to be very significant. Imagine Node.js creating a Swoole table and PHP being able to access it (or vice versa) without a cost in performance. Or a Java package that can access Swoole Channel objects created within PHP. This is possible with the route that Swoole is on. It can be done right now if someone endeavored to do it.

**Comment by embluk** [comment]
I disagree with this structure, keeping everything the way it is is much simpler and easy to manage when running Swoole. Some parts cannot be done in PHP and if they could, it would be much slower. I like the way Swoole works currently.

I think, features like middleware, routes and database ORMs etc. need to be left to frameworks created for Swoole; just as I am doing with a dedicated web HTTP framework for Swoole https://github.com/Superbition/Phase

Swoole should focus on what they started which is being a networking framework for PHP.

**Comment by matyhtf** [comment]
Reference https://github.com/swoole/swoole-cli  

The future development direction of Swoole is to become a standalone executable program, rather than just a PHP extension. Similar to the relationship between `Node.js` and `V8`, we use `Zend PHP` as the programming language and virtual machine, but do not use its runtime design such as `PHP-FPM`.
